gremlin-dotnet/src/Gremlin.Net/Driver/Connection.cs

```diff
@@ -68,18 +76,53 @@ public async Task ConnectAsync()
 
         public async Task CloseAsync()
         {
+            if (_sessionEnabled)
+            {
+                await CloseSession().ConfigureAwait(false);
+            }
             await _webSocketConnection.CloseAsync().ConfigureAwait(false);
         }
 
+        private async Task CloseSession()
+        {
+            // build a request to close this session, 'gremlin' in args as tips and not be executed actually
+            var msg = RequestMessage.Build(Tokens.OpsClose).Processor(Tokens.ProcessorSession)
+              .AddArgument(Tokens.ArgsGremlin, "session.close()").Create();
+
+            await SendAsync(msg).ConfigureAwait(false);
+        }
+
         public bool IsOpen => _webSocketConnection.IsOpen;
 
         private async Task SendAsync(RequestMessage message)
         {
-            var graphsonMsg = _graphSONWriter.WriteObject(message);
+            var msg = message;
```

Comment:
(nitpick) This shouldn't be necessary as you can also assign `message` again below:

```c#
message = RebuildSessionMessage(message);
```

Reply:
thanks, got it